pageserver/src/tenant/storage_layer.rs

```diff
@@ -336,6 +311,952 @@ impl LayerAccessStats {
+            inner: Default::default(),
+            version: AtomicUsize::new(0),
+            #[cfg(test)]
+            evicted: tokio::sync::Notify::default(),
+        }
+    }
+
+    pub(crate) fn for_written(
+        conf: &'static PageServerConf,
+        timeline: &Arc<Timeline>,
+        desc: PersistentLayerDesc,
+    ) -> anyhow::Result<ResidentLayer> {
+        let path = conf
+            .timeline_path(&desc.tenant_id, &desc.timeline_id)
+            .join(desc.filename().to_string());
+
+        let mut resident = None;
+
+        let outer = Arc::new_cyclic(|owner| {
+            let inner = Arc::new(DownloadedLayer {
+                owner: owner.clone(),
+                kind: tokio::sync::OnceCell::default(),
+            });
+            resident = Some(inner.clone());
+            LayerE {
+                conf,
+                path,
+                desc,
+                timeline: Arc::downgrade(timeline),
+                have_remote_client: timeline.remote_client.is_some(),
+                access_stats: LayerAccessStats::empty_will_record_residence_event_later(),
+                wanted_garbage_collected: AtomicBool::new(false),
+                wanted_evicted: AtomicBool::new(false),
+                inner: tokio::sync::Mutex::new(Some(ResidentOrWantedEvicted::Resident(inner))),
+                version: AtomicUsize::new(0),
+                #[cfg(test)]
+                evicted: tokio::sync::Notify::default(),
+            }
+        });
+
+        // FIXME: ugly, but if we don't do this check here, any error will pop up at read time
+        // but we cannot check it because DeltaLayerWriter and ImageLayerWriter create the
+        // instances *before* renaming the file to final destination
+        // anyhow::ensure!(
+        //     outer.needs_download_blocking()?.is_none(),
+        //     "should not need downloading if it was just written"
+        // );
+
+        // FIXME: because we can now do garbage collection on drop, should we mark these files as
+        // garbage collected until they get really get added to LayerMap? consider that files are
+        // written out to disk, fsynced, renamed by `{Delta,Image}LayerWriter`, then waiting for
+        // remaining files to be generated (compaction, create_image_layers) before being added to
+        // LayerMap. We could panic or just error out during that time, even for unrelated reasons,
+        // but the files would be left.
+
+        Ok(ResidentLayer {
+            _downloaded: resident.expect("just wrote Some"),
+            owner: outer,
+        })
+    }
+
+    /// Evict the the layer file as soon as possible, but then allow redownloads to happen.
+    pub(crate) async fn evict(
+        &self,
+        _: &RemoteTimelineClient,
+    ) -> Result<bool, super::timeline::EvictionError> {
+        assert!(
+            self.have_remote_client,
+            "refusing to evict without a remote timeline client"
+        );
```

Comment:
What would this give? Protection against malicious callers? :)

Reply:
IMO duplication of such state is asking for future bugs, your call. Resolve if you don't think it makes sense.

Reply:
Now there's a witness and non-option RemoteTimelineClient. I don't think you can have a None-ish remote timeline client... Well, I guess it could be stopped but still it's configured.